src/libraries/Microsoft.PowerFx.Core/Types/DefinedTypeDependencyGraph.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Syntax.Visitors;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Types
+{
+    internal class DefinedTypeDependencyGraph
+    {
+        private readonly IEnumerable<DefinedType> _definedTypes;
+        private readonly ReadOnlySymbolTable _globalTypes;
+
+        // stores unresolvedType -> its dependencies
+        private readonly Dictionary<DefinedType, HashSet<string>> _typeWithDependency;
+
+        // stores unresolvedType -> its dependendents
+        private readonly Dictionary<string, HashSet<DefinedType>> _invertedDependency;
+
+        // queue with types ready to be resolved
+        private readonly Queue<DefinedType> _tsQueue;
+
+        private readonly SymbolTable _definedTypeSymbolTable;
+
+        // TODO: Add more future type names
```

Comment:
>TODO [](http://example.com/codeflow?start=11&length=4)

Is this a TODO before shipping? Hard to restrict things post shipping.

Reply:
before shipping, seeking help from reviewers on what all to restrict